aptos-move/move-examples/veiled_coin/sources/veiled_coin.move

```diff
@@ -0,0 +1,376 @@
+    const MAX_BITS_IN_VALUE : u64 = 64;
+
+    /// The domain separation tag (DST) used for the Bulletproofs prover.
+    const VEILED_COIN_DST : vector<u8> = b"AptosVeiledCoinExample";
+
+    //
+    // Core data structures.
+    //
+
+    /// Main structure representing a coin in an account's custody.
+    struct VeiledCoin<phantom CoinType> {
+        /// Pedersen commitment (under the default Bulletproofs CK) to a number of coins v \in [0, 2^{64}), an invariant
+        /// that is enforced throughout the code
+        private_value: Commitment,
+    }
+
+    /// A holder of a specific coin types and associated event handles.
+    /// These are kept in a single resource to ensure locality of data.
+    struct VeiledCoinStore<phantom CoinType> has key {
+        /// A Pedersen commitment to a value v \in [0, 2^{64}), an invariant that is enforced throughout the code.
+        private_balance: CompressedRistretto,
+        deposit_events: EventHandle<DepositEvent>,
+        withdraw_events: EventHandle<WithdrawEvent>,
+    }
+
+    /// Holds a signer capability for the resource account created when initializing this module. This account houses a
+    /// coin::CoinStore<T> for every type of coin T that is veiled.
+    struct VeiledCoinMinter has store, key {
+        signer_cap: account::SignerCapability,
+    }
+
+    /// Event emitted when some amount of a coin is deposited into an account.
+    struct DepositEvent has drop, store {
+        // We cannot leak any information about how much has been deposited.
+    }
+
+    /// Event emitted when some amount of a coin is withdrawn from an account.
+    struct WithdrawEvent has drop, store {
+        // We cannot leak any information about how much has been withdrawn.
+    }
+
+    //
+    // Module initialization, done only once when this module is first published on the blockchain
+    //
+
+    /// Initializes a so-called "resource" account which will maintain a coin::CoinStore<T> resource for all Coin<T>'s
+    /// that have been converted into a VeiledCoin<T>.
+    fun init_module(deployer: &signer) {
+        assert!(bulletproofs::get_max_range_bits() >= 64, ERANGE_PROOF_SYSTEM_HAS_INSUFFICIENT_RANGE);
+
+        // Create the resource account. This will allow this module to later obtain a `signer` for this account and
+        // transfer Coin<T>'s into its CoinStore<T> before minting a VeiledCoin<T>.
+        let (_resource, signer_cap) = account::create_resource_account(deployer, vector::empty());
+
+        move_to(deployer,
+            VeiledCoinMinter {
+                signer_cap
+            }
+        )
+    }
+
+    //
+    // Entry functions
+    //
+
+    /// Takes `amount` of `coin::Coin<CoinType>` coins from `sender`, wraps them inside a `VeiledCoin<CoinType>` and
+    /// sends them back to `sender`.
+    public entry fun mint<CoinType>(sender: &signer, amount: u64) acquires VeiledCoinMinter, VeiledCoinStore {
```

Comment:
A better name for this might be `veil`

Reply:
Oh, not called `veil` because it does not use any randomness in `mint_from_coin`

Reply:
`wrap`?